File: android/src/main/java/io/parity/signer/components/networkicon/jdenticon/jdenticon_kotlin/SvgRenderer.kt
package io.parity.signer.components.networkicon.jdenticon.jdenticon_kotlin

internal interface Renderer {
	fun setBackground(fillColor: String)
	fun beginShape(color: String)
	fun endShape()
	fun addPolygon(points: List<Point>)
	fun addCircle(point: Point, diameter: Float, counterClockwise: Boolean)
	fun finish()
}
internal class SvgRenderer(target: SvgWriter) : Renderer {

    var _pathsByColor = HashMap<String, SvgPath>()
    var _target = target
    var size = target.size
    lateinit var _path: SvgPath

    override fun setBackground(fillColor: String) {
        var re = Regex("^(#......)(..)?")
        var match = re.matchEntire(fillColor)?.groups?.get(1)?.value
        var opacityMatch = re.matchEntire(fillColor)?.groups?.get(2)?.value
        var opacity = opacityMatch?.let {
            opacityMatch.toInt(16) / 255f
        } ?: 1f
        var colorMatch = re.matchEntire(fillColor)?.groups?.get(1)?.value
        var color = colorMatch?.let {
            colorMatch
        } ?: "000000"
        this._target.setBackground(color, opacity)
    }

    override fun beginShape(color: String) {
        if (this._pathsByColor[color] == null) {
            this._pathsByColor[color] = SvgPath()
        }
        this._path = this._pathsByColor[color]!!
    }

    override fun endShape() {

    }

    override fun addPolygon(points: List<Point>) {
        this._path.addPolygon(points)
    }

    override fun addCircle(point: Point, diameter: Float, counterClockwise: Boolean) {
        this._path.addCircle(point, diameter, counterClockwise)
    }

    override fun finish() {
        /*
            Set is not ordered, so we can have different order of _pathsByColor.keys
            across different versions (for example, kotlin-js and kotlin-jvm iterate this set
            in a different order). To make tests stable we have to
            sort keys. It does not affect performance alot, because there are
            pretty small amount of keys.
        */
        for (color in this._pathsByColor.keys.sorted()) {
            this._target.append(color, this._pathsByColor[color]!!.dataString)
        }
    }
}

File: ios/Packages/Jdenticon/Sources/Jdenticon/Svg/SvgRenderer.swift
//
//  SvgRenderer.swift
//
//
//  Created by Krzysztof Rodak on 22/08/2023.
//

import Foundation

protocol Renderer {
    func setBackground(fillColor: String)
    func beginShape(color: String)
    func endShape()
    func addPolygon(points: [CGPoint])
    func addCircle(point: CGPoint, diameter: CGFloat, counterClockwise: Bool)
    func finish()
}

class SvgRenderer: Renderer {
    private var _pathsByColor = [String: SvgPath]()
    private var _target: SvgWriter
    private var size: Int {
        _target.size
    }

    private var _path: SvgPath?

    init(target: SvgWriter) {
        _target = target
    }

    func setBackground(fillColor: String) {
        let re = try! NSRegularExpression(pattern: "^(#......)(..)?")
        if let match = re.firstMatch(in: fillColor, range: NSRange(fillColor.startIndex..., in: fillColor)),
           let colorMatchRange = Range(match.range(at: 1), in: fillColor),
           let opacityMatchRange = Range(match.range(at: 2), in: fillColor) {
            let colorMatch = String(fillColor[colorMatchRange])
            let opacityMatch = String(fillColor[opacityMatchRange])
            let opacity = CGFloat(UInt32(opacityMatch, radix: 16) ?? 0) / 255.0
            _target.setBackground(fillColor: colorMatch, opacity: opacity)
        }
    }

    func beginShape(color: String) {
        if _pathsByColor[color] == nil {
            _pathsByColor[color] = SvgPath()
        }
        _path = _pathsByColor[color]
    }

    func endShape() {
        _path = nil
    }

    func addPolygon(points: [CGPoint]) {
        _path?.addPolygon(points: points)
    }

    func addCircle(point: CGPoint, diameter: CGFloat, counterClockwise: Bool) {
        _path?.addCircle(point: point, diameter: diameter, counterClockwise: counterClockwise)
    }

    func finish() {
        _pathsByColor.keys.sorted().forEach { color in
            if let dataString = _pathsByColor[color]?.dataString {
                _target.append(color: color, dataString: dataString)
            }
        }
    }
}
